bentoml/yatai/deployment/azure_function/operator.py

```diff
@@ -0,0 +1,724 @@
+import re
+import shutil
+import subprocess
+import logging
+import sys
+
+import docker
+
+from bentoml import config
+from bentoml.utils.tempdir import TempDirectory
+from bentoml.saved_bundle import loader
+from bentoml.yatai.deployment.azure_function.templates import (
+    AZURE_FUNCTION_HOST_JSON,
+    AZURE_FUNCTION_LOCAL_SETTING_JSON,
+    AZURE_API_INIT_PY,
+    AZURE_API_FUNCTION_JSON,
+    BENTO_SERVICE_AZURE_FUNCTION_DOCKERFILE,
+)
+from bentoml.yatai.deployment.operator import DeploymentOperatorBase
+from bentoml.yatai.deployment.utils import ensure_docker_available_or_raise
+from bentoml.exceptions import (
+    BentoMLException,
+    MissingDependencyException,
+    AzureServiceError,
+    YataiDeploymentException,
+)
+from bentoml.yatai.proto.deployment_pb2 import (
+    ApplyDeploymentResponse,
+    DeploymentState,
+    DescribeDeploymentResponse,
+    DeleteDeploymentResponse,
+)
+from bentoml.yatai.proto.repository_pb2 import GetBentoRequest
+from bentoml.yatai.status import Status
+from bentoml.configuration import LAST_PYPI_RELEASE_VERSION
+
+logger = logging.getLogger(__name__)
+
+
+def ensure_azure_cli_available_and_login_or_raise():
+    try:
+        account_info = _call_az_cli(
+            command=['az', 'account', 'show'], message='show Azure account'
+        )
+        if not account_info['user'] or not account_info['homeTenantId']:
+            # https://docs.microsoft.com/en-us/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest
+            # Attempt to login with azure service principal values from bentoml
+            # configuration.
+            service_principal_app_id = config.get('yatai_service').get(
+                'azure_service_principal_app_url', None
+            )
+            service_principal_cert = config('yatai_service').get(
+                'azure_service_principal_cert', None
+            )
+            service_principal_tenant = config('yatai_service').get(
+                'azure_service_principal_tenant', None
+            )
+            if (
+                service_principal_app_id
+                and service_principal_cert
+                and service_principal_tenant
+            ):
+                service_principal_login_result = _call_az_cli(
+                    command=[
+                        'az',
+                        'login',
+                        '--service-principal',
+                        '--user-name',
+                        service_principal_app_id,
+                        '--password',
+                        service_principal_cert,
+                        '--tenant',
+                        service_principal_tenant,
+                    ],
+                    message='login Azure CLI with service principal account',
+                )
+                if service_principal_login_result['state'] != 'Enabled':
+                    raise BentoMLException(
+                        'Azure service principal account is not enabled'
```

Comment:
Improve the error message here?  somethign like - "Failed to login Azure Service account with {app_id}, {cert_file_path}, {tenant}. CLI response: {service_principal_login_result}"